File: Assign4/task1.py
from pyspark import SparkContext
import sys
import itertools
import os
from pyspark.sql import SQLContext
from graphframes import GraphFrame
import time
from pyspark import SparkConf

start = time.time()
os.environ["PYSPARK_SUBMIT_ARGS"] = ("--packages graphframes:graphframes:0.6.0-spark2.3-s_2.11")

conf=SparkConf()
#conf.set("spark.executor.memory", "4g")
#conf.set("spark.driver.memory", "4g")
sc = SparkContext('local[3]').getOrCreate(conf)

sqlContext = SQLContext(sc)

sc.setLogLevel("ERROR") 
sample_text = sc.textFile(sys.argv[2]).map(lambda x : x.split(','))
sample_first = sample_text.take(1) 
sample_text = sample_text.filter(lambda x : x !=sample_first[0]).map(lambda x : (x[1],[x[0]])).reduceByKey(lambda x,y: x+y)
#edges = sample_text.map(lambda x : (x[0],[x[1])])).reduceByKey(lambda x,y : x+y).map(lambda x :(1,[x])).reduceByKey(lambda x,y : x+y).flatMap(lambda x : itertools.combinations(x[1],2)).filter(lambda x : len(set(x[0][1]).intersection(set(x[1][1]))) >= 7).map(lambda x : (x[0][0].encode('UTF-8'),x[1][0].encode('UTF-8'))).distinct().map(lambda x : [x[0],x[1]])

t = int(sys.argv[1])
edges = sample_text.flatMap(lambda x : ([x[0],y] for y in list(itertools.permutations(x[1],2)))).filter(lambda x : len(set(x[1]))  != 1).map(lambda x : (x[1],[x[0]])).reduceByKey(lambda x,y : x+y).filter(lambda x : len(set(x[1])) >= t).map(lambda x :(x[0][0],x[0][1])).distinct().map(lambda x :[x[0],x[1]])
edgesDF = sqlContext.createDataFrame(edges,["src","dst"])
vertices = edges.flatMap(lambda x: ((x[0]),(x[1]))).distinct().map(lambda x : [x])

verticesDF = sqlContext.createDataFrame(vertices,["id"])

g = GraphFrame(verticesDF, edgesDF)
comm = g.labelPropagation(maxIter=5).select('label','id')
comm_rdd = comm.rdd.map(lambda x:(x[0],[x[1]])).reduceByKey(lambda x,y: x+y).sortBy(lambda x : (len(x[1]),sorted(x[1]))).collect()

f = open(sys.argv[3],"w")
for community in comm_rdd :  
    print(community)
    f.write(str(sorted(community[1])).strip("[]"))
    f.write("\n")
f.close()
sc.stop()
print("Duration : " + str(time.time()-start))

File: Assign4/task1.scala
import java.io.{BufferedWriter, File, FileWriter}

import org.apache.spark.SparkContext
import org.apache.spark.SparkConf
import org.apache.spark.sql.SQLContext
import org.apache.spark.sql.types.StructType
import org.graphframes.GraphFrame
import org.json4s.JsonAST.JString

import scala.math.Ordering.Implicits.seqDerivedOrdering


object task1 {

    def main(args: Array[String]): Unit = {
      val conf = new SparkConf().setAppName("spark-test").setMaster("local[*]")
      val sc = new SparkContext(conf)
      val sqlContext = new SQLContext(sc)
      sc.setLogLevel("ERROR")
      val threshold = args(0).trim().toInt
      val file = args(2)
      val textFile = sc.textFile(args(1).trim()).map(x => x.split(","))
      val header = textFile.take(1)
      val filter = textFile.filter(x => x != header.toList(0))
      val sample_text = filter.map(x => (x(1), x(0))).reduceByKey((x,y)=>x+" "+y).map(x=>(x._1,x._2.split(" ")))
      val edges = sample_text.flatMap(x =>( for(y <-x._2.permutations) yield (Tuple1(y), x._1)  ))
        //.reduceByKey((x,y) => x + " " + y).filter(x =>  x._2.split(" ").toSet.toList.length >= threshold).map(x => (x._1)).distinct()
      println(edges.take(1).toList)
      println("yes")
            val edgesDF = sqlContext.createDataFrame(edges.collect()).toDF("src","dst")

            val vertices = edges.map( x=> Tuple1(x._1)).distinct()

            val verticesDF =sqlContext.createDataFrame(vertices).toDF("id")

            val g = GraphFrame( verticesDF, edgesDF)

            val comm = g.labelPropagation.maxIter(5).run()

            val comm_rdd = comm.select("label","id").rdd.map( x=>(JString(x(0).toString),x(1).toString)).reduceByKey((x, y)=>x+", "+y).map(x=>(x._1,x._2.split(", ").sorted)).sortBy( x =>(x._2.toList.length,x._2.toList.sorted)).collect()

            val bw = new BufferedWriter(new FileWriter(file))

              for ( community <-comm_rdd ) {
                var i = 0
                for (node<- community._2){
                  if(i==0) {
                    bw.write("'"+node+"'")
                    i = i + 1

                  }else{
                    bw.write(", '"+node+"'")
                  }
                }

            bw.write("\n")
              }
            bw.close()

        }

}

